### DISCUSSION TITLE
A global `alert()` equivalent

### DISCUSSION THREAD

**Discussion by Heartade** [ariakit/ariakit#1706] [Questions] [answered]
For now, when I need to alert with a message while the user is doing something on a modal, I've been using nested dialogs. However, a `Dialog` component for every alert in an app feels a bit too repetitive. I've tried putting the alert dialog component on the app root and exposing its `DialogState` to a `Context`, but the dialog isn't interactable as (I guess) the focus is trapped elsewhere if a modal `Dialog` is already open. Making all dialogs non-modals wouldn't be a good idea either. Is there any way to kind of 'hijack' the focus from an already open modal?

**Comment by diegohaz** [comment, marked as answer]
Hey @Heartade! Could you put together a CodeSandbox demonstrating what you mean so we can iterate on it?

**Reply by Heartade** [reply]
Sure! Here is a rough representation of what I'm trying to do. https://codesandbox.io/s/reakit-global-alert-focus-trap-m1wor4?file=/src/App.tsx

**Reply by Heartade** [reply]
Another thing to note is that in the actual project some of these modals represent required actions in our app, which means 'hideOnInteractOutside={false}' and 'hideOnEsc={false}'.

**Reply by diegohaz** [reply]
Thank you for the CodeSandbox. You can temporarily disable `modal` while the alert is open:

```jsx
<Dialog modal={!alert.open} portal />
```

[CodeSandbox](https://codesandbox.io/s/ariakit-global-dialog-njcleq?file=/src/App.tsx:381-420)

`portal` is necessary there since `modal={false}` would default to `portal={false}`. You can abstract it into a separate component, so you don't need to repeat this logic every time.

This is similar to what the library does internally for nested dialogs. Some parent dialog behaviors are disabled while the nested dialog is open.

**Reply by diegohaz** [reply]
Also, for the cases where you need `hideOnInteractOutside={true}` (the default), you can pass a function to ignore the alert dialog:
```jsx
<Dialog
  hideOnInteractOutside={(event) => {
    if (alert.contentElement?.contains(event.target)) return false;
    return true;
  }}
/>
```

That's also what the library does for nested dialogs (when they're nested in the JSX tree):

https://github.com/ariakit/ariakit/blob/32a61f7f691089788e85dbce35bcccd5878b2bd9/packages/ariakit/src/dialog/__utils/use-hide-on-interact-outside.ts#L91-L92

**Reply by Heartade** [reply]
Thanks so much! Had no idea I could pass a function to hideOnInteractOutside.

**Reply by diegohaz** [reply]
Yeah! Most of the `doSomethingOnEvent` props accept a function in addition to a boolean. We made this decision a few months ago, so not all props are converted yet.
